File: apps/backend/app/schemas/transfer.py
"""Pydantic schemas for transfer-related operations."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, ConfigDict


class TransferResponse(BaseModel):
    """Schema representing a transfer job returned by the API."""

    id: int
    file_id: Optional[int]
    type: str
    status: str
    progress: Optional[float] = Field(None, description="Completion percentage (0.0–1.0).")
    bytes_transferred: Optional[int] = Field(
        None,
        description="Number of bytes transferred so far.",
    )
    total_bytes: Optional[int] = Field(
        None,
        description="Total number of bytes for the transfer.",
    )
    started_at: datetime
    last_progress_at: Optional[datetime] = None
    current_speed_bps: Optional[float] = None
    average_speed_bps: Optional[float] = None
    eta_seconds: Optional[int] = None
    priority: Optional[int] = None
    queue_position: Optional[int] = None
    pause_until: Optional[datetime] = Field(
        None,
        description="If set, the transfer is paused until this UTC time.",
    )
    created_at: datetime
    updated_at: datetime

    # Enable ORM mode using the new Pydantic v2 configuration API.
    model_config = ConfigDict(from_attributes=True)

File: apps/desktop/final_repo_updated/repo/apps/desktop/src/types/transfer.ts
/**
 * Transfer jobs track uploads and downloads of files. The backend
 * exposes detailed progress metrics such as speed and ETA. Null
 * values indicate that information is not yet available for a job.
 */
export interface Transfer {
  id: number;
  file_id: number | null;
  type: string;
  status: string;
  progress: number | null;
  bytes_transferred: number | null;
  total_bytes: number | null;
  started_at: string;
  last_progress_at: string | null;
  current_speed_bps: number | null;
  average_speed_bps: number | null;
  eta_seconds: number | null;
  priority: number | null;
  queue_position: number | null;
  pause_until: string | null;
  created_at: string;
  updated_at: string;
}
